apps/src/code-studio/reporting.js

```diff
@@ -244,9 +244,22 @@ reporting.sendReport = function (report) {
   // Post milestone iff the server tells us.
   // Check a second switch if we passed the last level of the script.
   // Keep this logic in sync with ActivitiesController#milestone on the server.
-  if (appOptions.postMilestone ||
-    (appOptions.postFinalMilestone && report.pass && appOptions.level.final_level)) {
+  var postMilestone = appOptions.postMilestone;
+  // Don't do this milestone post if postFailedRunMilestone is enabled,
```

Comment:
Should this say `if postFailedRunMilestone is disabled`?

It might be worth a singular comment somewhere describing how these different settings interact (`postMilestone/postFinalMilestone/postFailedRunMilestone`).

i.e. It's not immediately clear to me what happens with some combinations of these. Like is postFinalMilestone=true meant to be used with postMilestone=false so that we only post milestones on the last level (I think this is the case), or is it meant to be used as postMilestone=true and postFinalMilestone=false so that we post everywhere but the last level.

It's not immediately clear what the order of precedence is for these settings.

Reply:
Might actually be worth a small helper function that takes all the relevant settings as an input and output a boolean. Then we could also write a set of unit tests to document/validate expectations.